pkg/client/command_line.go

```diff
@@ -15,62 +16,215 @@ func AddArmadaApiConnectionCommandlineArgs(rootCmd *cobra.Command) {
-	err = viper.MergeInConfig()
-
-	if err != nil {
-		switch err.(type) {
-		case viper.ConfigFileNotFoundError:
-			// This only occurs when looking for the default .armadactl file and it is not present
-			// This is not an error as users don't have to specify it, so do nothing
-		default:
-			log.Errorf("Can't read config file %s because %s\n", viper.ConfigFileUsed(), err)
-			os.Exit(1)
+	// merge in new config with those loaded from armadactl-defaults.yaml
+	// (note the call to viper.MergeInConfig instead of viper.ReadInConfig)
+	if err := viper.MergeInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
+			// Config file not found; ignore
+		} else {
+			return fmt.Errorf("[LoadCommandlineArgsFromConfigFile] error reading config file %s: %s", viper.ConfigFileUsed(), err)
 		}
 	}
+
+	return nil
 }
 
+// ExtractCommandlineArmadaApiConnectionDetails extracts Armada server connection details from the
+// config loaded into viper. Hence, this function must be called after loading config into viper,
+// e.g., by calling LoadCommandlineArgsFromConfigFile.
 func ExtractCommandlineArmadaApiConnectionDetails() *ApiConnectionDetails {
 	apiConnectionDetails := &ApiConnectionDetails{}
 	viper.Unmarshal(apiConnectionDetails)
 	return apiConnectionDetails
 }
+
+// ConfigYamlString returns a string representation of the merged configuration stored in viper.
+func ConfigYamlString() (string, error) {
+	c := viper.AllSettings()
+	bs, err := yaml.Marshal(c)
+	if err != nil {
+		return "", fmt.Errorf("[ConfigYamlString] unable to marshal config to yaml: %s", err)
+	}
+	return string(bs), nil
+}
+
+type armadaClientConfig struct {
+	ApiVersion     string
+	Users          map[string]*userConfig
+	Clusters       map[string]*clusterConfig
+	Contexts       map[string]*contextConfig
+	CurrentContext string
+}
+
+type userConfig struct {
+	Username string
+	Password string
+}
+
+type clusterConfig struct {
+	Server                string
+	InsecureSkipTlsVerify bool
+}
+
+type contextConfig struct {
+	Cluster string
+	User    string
+}
+
+// validate checks that no names are the empty string, that users/clusters references by the
+// contexts exists, and that the current context is defined. Other methods on armadaClientConfig
+// are themselves responsible for performing any further validation.
+func (config *armadaClientConfig) validate() error {
+	if config == nil {
+		return fmt.Errorf("[armadaClientConfig.validate] config is nil")
+	}
+
+	if config.Users != nil {
+		for name, user := range config.Users {
+			if name == "" {
+				return fmt.Errorf("[armadaClientConfig.validate] empty user names not allowed")
+			}
+			if user == nil {
+				return fmt.Errorf("[armadaClientConfig.validate] user %s is nil", name)
+			}
+		}
+	}
+
+	if config.Clusters != nil {
+		for name, cluster := range config.Clusters {
+			if name == "" {
+				return fmt.Errorf("[armadaClientConfig.validate] empty cluster names not allowed")
```

Comment:
I'm not familiar with go logging, but in logging frameworks in other languages you wouldn't need to add all the `[armadaClientConfig.Validate]`s because the logging framework could add it from the logging context.  It may be that something is possible here.  If not I'd be tempted to make it a const string and add it via the format string, but I accept that's really a matter of taste!

Reply:
:)